flink-connectors/flink-connector-elasticsearch7/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBuilder.java

```diff
@@ -0,0 +1,276 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.elasticsearch.sink;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.connector.base.DeliveryGuarantee;
+import org.apache.flink.streaming.connectors.elasticsearch.ElasticsearchSinkBase;
+import org.apache.flink.util.InstantiationUtil;
+
+import org.apache.http.HttpHost;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Builder to construct a {@link ElasticsearchSink}.
+ *
+ * <p>The following example shows the minimum setup to create a ElasticsearchSink that submits
+ * actions on checkpoint.
+ *
+ * <pre>{@code
+ * Elasticsearch<String> sink = Elasticsearch
+ *     .builder()
+ *     .setHosts(MY_ELASTICSEARCH_HOSTS)
+ *     .setProcessor(MY_ELASTICSEARCH_PROCESSOR)
+ *     .setDeliveryGuarantee(DeliveryGuarantee.AT_LEAST_ONCE)
+ *     .build();
+ * }</pre>
+ *
+ * @param <IN> type of the records converted to Elasticsearch actions
+ */
+@PublicEvolving
+public class ElasticsearchSinkBuilder<IN> {
+
+    private int bulkFlushMaxActions = -1;
+    private int bulkFlushMaxMb = -1;
+    private long bulkFlushInterval = -1;
+    private ElasticsearchSinkBase.FlushBackoffType bulkFlushBackoffType;
+    private int bulkFlushBackoffRetries = -1;
+    private long bulkFlushBackOffDelay = -1;
+    private DeliveryGuarantee deliveryGuarantee = DeliveryGuarantee.NONE;
+    private List<HttpHost> hosts;
+    private ElasticsearchProcessor<? extends IN> processor;
+    private String username;
+    private String password;
+    private String connectionPathPrefix;
+
+    ElasticsearchSinkBuilder() {}
+
+    /**
+     * Sets the hosts where the Elasticsearch cluster nodes are reachable.
+     *
+     * @param hosts http addresses describing the node locations
+     */
+    public ElasticsearchSinkBuilder<IN> setHosts(HttpHost... hosts) {
+        checkNotNull(hosts);
+        checkState(hosts.length > 0, "Hosts cannot be empty.");
+        this.hosts = Arrays.asList(hosts);
+        return this;
+    }
+
+    /**
+     * Sets the processor which is invoked on every record to convert it to Elasticsearch actions.
+     *
+     * @param processor to process records into Elasticsearch actions.
+     * @return {@link ElasticsearchSinkBuilder}
+     */
+    public <T extends IN> ElasticsearchSinkBuilder<T> setProcessor(
+            ElasticsearchProcessor<? extends T> processor) {
```

Comment:
In a way the processor is like the `SerializationSchema`. How do they relate? Should we add some Processor builder similar to `KafkaSerializationSchema`?

Reply:
> In a way the processor is like the `SerializationSchema`. How do they relate? Should we add some Processor builder similar to `KafkaSerializationSchema`?

You are right, I was also wondering about the relation between the usually used `SerializationSchema`s and this elasticsearch sink way. So far I do not see a comfortable way to offer a nice build.

Reply:
Let's create a follow-up on that. I feel that this is again a building block but I might be wrong because ES is not working on byte[] as other sinks.